lighthouse-core/audits/injected-iframes.js

```diff
@@ -0,0 +1,200 @@
+  /** Label for a column in a data table; entries will be the timestamp in milliseconds an event occurs. */
+  columnTimestamp: 'Timestamp',
+  /** Label for audit identifying the number of iframes likely affecting the page. */
+  displayValue: `{itemCount, plural,
+    =1 {1 iframe}
+    other {# iframes}
+    }`,
+  /** Description of a Lighthouse audit that tells the user potential causes of CLS. This is displayed after a user expands the section to see more. No character length limits. 'Learn More' becomes link text to additional documentation. */
+  description: 'Injecting an iframe with a correctly sized container can reduce layout shifting and improve CLS. [Learn More](https://web.dev/optimize-cls/#ads-embeds-and-iframes-without-dimensions)',
+};
+
+const str_ = i18n.createMessageInstanceIdFn(__filename, UIStrings);
+
+/**
+ * LayoutShift timing windows are determined by trace events. We consider a timing window to be 
+ * between a ScheduleStyleRecalculation event and the end of its possibly corresponding UpdateLayerTree event.
+ * We assume that an UpdateLayerTree event belongs to a ScheduleStyleRecalculation event if it occurs within
+ * 20ms. This window is then determined to be a "layoutShiftWindow", if a LayoutShift event occurs
+ * within a time frame of this UpdateLayerTree event.
+ * @param {{ event: LH.TraceEvent; timing: any; duration: number; }[]} layoutEvents
+ * @return {Array<LH.Artifacts.DOMWindow>}
+ */
+function getLayoutShiftWindows(layoutEvents) {
+  /**@type {Array<LH.Artifacts.DOMWindow>} */
+  const shiftWindows = [];
+
+  // TODO: use a map for this ?
+  for (let i = 0; i < layoutEvents.length; i++) {
+    const event = layoutEvents[i].event;
+    const SSRStart = layoutEvents[i].timing;
+    const SSREnd = SSRStart + layoutEvents[i].duration;
+    // look for a ScheduleStyleRecalculation
+    if (event.name === 'ScheduleStyleRecalculation') {
+      // look for a ULT within this limit, assume that a ULT belongs to a SSR if occurs within 20ms
+      const limit = SSRStart +  20;
+      for (let j = i+1; j < layoutEvents.length; j++) {
+        const ULTStart = layoutEvents[j].timing;
+        const ULTEnd = ULTStart + layoutEvents[j].duration;
+        if (layoutEvents[j].event.name === 'UpdateLayerTree' && ULTStart >= SSRStart && ULTStart < limit) {
+          // Look for ULT's layout shift if any
+          for(let k = j+1; k < layoutEvents.length; k++) {
+            // If there is a layout shift within an update layer tree, it may cause a layout shift.
+            if (layoutEvents[k].event.name === 'LayoutShift') {
+              const layoutShiftStart = layoutEvents[k].timing;
+              const layoutShiftEnd = layoutShiftStart + layoutEvents[k].duration;
+              if (layoutShiftStart >= ULTStart && layoutShiftStart <= ULTEnd) {
+                shiftWindows.push({start: SSRStart, end: ULTEnd});
+              }
+            }
+          }
+        }
+      }
+    }
```

Comment:
Are this windows ever nested? Wondering if we should do `i = j;` at the end here to speed things along every time a window is found.